server/snapshot/snapshot.go

```diff
@@ -0,0 +1,161 @@
+package snapshot
+
+import (
+	"bytes"
+	"encoding/json"
+	"github.com/Sirupsen/logrus"
+
+	"github.com/docker/notary/server/storage"
+	"github.com/docker/notary/tuf/data"
+	"github.com/docker/notary/tuf/signed"
+)
+
+// GetOrCreateSnapshotKey either creates a new snapshot key, or returns
+// the existing one. Only the PublicKey is returned. The private part
+// is held by the CryptoService.
+func GetOrCreateSnapshotKey(gun string, store storage.KeyStore, crypto signed.CryptoService, createAlgorithm string) (data.PublicKey, error) {
+	keyAlgorithm, public, err := store.GetKey(gun, data.CanonicalSnapshotRole)
+	if err == nil {
+		return data.NewPublicKey(keyAlgorithm, public), nil
+	}
+
+	if _, ok := err.(*storage.ErrNoKey); ok {
+		key, err := crypto.Create("snapshot", createAlgorithm)
+		if err != nil {
+			return nil, err
+		}
+		logrus.Debug("Creating new snapshot key for ", gun, ". With algo: ", key.Algorithm())
+		err = store.SetKey(gun, data.CanonicalSnapshotRole, key.Algorithm(), key.Public())
+		if err == nil {
+			return key, nil
+		}
+
+		if _, ok := err.(*storage.ErrKeyExists); ok {
+			keyAlgorithm, public, err = store.GetKey(gun, data.CanonicalSnapshotRole)
+			if err != nil {
+				return nil, err
+			}
+			return data.NewPublicKey(keyAlgorithm, public), nil
+		}
+		return nil, err
+	}
+	return nil, err
+}
+
+// GetOrCreateSnapshot either returns the exisiting latest snapshot, or uses
+// whatever the most recent snapshot is to generate a new one.
+func GetOrCreateSnapshot(gun string, store storage.MetaStore, cryptoService signed.CryptoService) ([]byte, error) {
+
+	d, err := store.GetCurrent(gun, "snapshot")
+	if err != nil {
+		return nil, err
+	}
+
+	sn := &data.SignedSnapshot{}
+	if d != nil {
+		err := json.Unmarshal(d, sn)
+		if err != nil {
+			logrus.Error("Failed to unmarshal existing snapshot")
+			return nil, err
+		}
+		if !snapshotExpired(sn) && !contentExpired(gun, sn, store) {
```

Comment:
`if snapshotExpired(sn)`, `contentExpired` will not be called and `sn.Signed.Meta` will not be updated.